bbq/compiler/peephole_pass.go

```diff
@@ -0,0 +1,170 @@
+// bytecode shifts are also in order of their offset,
+// so we can keep a counter for cumulative shift and patch jumps in the target order
+// ASSUMPTION: peephole patterns should never match a jump instruction
+func (o *PeepholeInstructionStaticTypeOptimizer) patchJumps(optimized []opcode.Instruction) {
+	cumShift := 0
+	currentShiftIndex := 0
+	for _, jump := range o.jumps {
+		jumpTarget := int(opcode.JumpTarget(optimized[jump]))
+
+		// apply all shifts up to the current jump target
+		for currentShiftIndex < len(o.bytecodeShifts) && o.bytecodeShifts[currentShiftIndex].Offset < jumpTarget {
+			cumShift += o.bytecodeShifts[currentShiftIndex].Shift
+			currentShiftIndex++
+		}
+
+		// patch jump
+		newJumpTarget := jumpTarget + cumShift
+		PatchJump(&optimized, jump, uint16(newJumpTarget))
+	}
+}
+
+func (o *PeepholeInstructionStaticTypeOptimizer) OptimizeInstructions(instructions []opcode.Instruction) []opcode.Instruction {
+	// collect all jump targets
+	for _, instruction := range instructions {
+		if opcode.IsJump(instruction) {
+			o.jumpTargets[int(opcode.JumpTarget(instruction))] = struct{}{}
+		}
+	}
+
+	optimized := make([]opcode.Instruction, 0, len(instructions))
+
+	for i := 0; i < len(instructions); i++ {
+		currentInstruction := instructions[i]
+		candidates := PatternsByOpcode[currentInstruction.Opcode()]
+
+		var matched bool
+		// check candidates for pattern match for each instruction
+		for _, candidate := range candidates {
+			candidateOpcodes := candidate.Opcodes
+			window := instructions[i : i+len(candidateOpcodes)]
+
+			if candidate.Match(window, i, o.jumpTargets) {
+				replacement := candidate.Replacement(window, o.compiler)
+				optimized = append(optimized, replacement...)
+
+				o.bytecodeShifts = append(o.bytecodeShifts, BytecodeShiftPair{
+					Offset: i,
+					// e.g. replace 2 with 1 -> shift is -1
+					Shift: len(replacement) - len(candidateOpcodes),
+				})
+
+				i += len(candidateOpcodes) - 1
+				matched = true
+				break
+			}
+		}
+		if !matched {
+			optimized = append(optimized, currentInstruction)
+
+			// if instruction is a jump, add it to the jumps list
+			// its important we do it in the optimized index, not the original index
+			// because its hard to calculate the optimized index from the original index later
+			if opcode.IsJump(currentInstruction) {
+				o.jumps = append(o.jumps, len(optimized)-1)
+			}
+		}
+	}
+
+	// Sort jumps by their target offsets
+	sort.Slice(o.jumps, func(a, b int) bool {
+		jumpATarget := opcode.JumpTarget(optimized[o.jumps[a]])
+		jumpBTarget := opcode.JumpTarget(optimized[o.jumps[b]])
+		// ascending order
+		return jumpATarget < jumpBTarget
+	})
+
+	o.patchJumps(optimized)
+
+	return optimized
+}
+func (o *PeepholeInstructionStaticTypeOptimizer) Optimize(code []opcode.Instruction) []opcode.Instruction {
+	o.bytecodeShifts = make([]BytecodeShiftPair, 0)
+	o.jumps = make([]int, 0)
+	o.jumpTargets = make(map[int]struct{})
```

Comment:
Move the allocation to the constructor function (later we should ideally allocate these lazily), and re-use the values here, instead of throwing away and re-allocating:

```suggestion
	o.bytecodeShifts = o.bytecodeShifts[:0]
	o.jumps = o.jumps[:0]
	clear(o.jumpTargets)
```